Native segwit support including basic BIP84 wallet.

AdamISZ opened the pull request:
* parsing of scripts and addresses of all segwit types.
* ability to verify arbitrary tx inputs for all segwit types.
* simplify mktx syntax (only allow [],[] args)
* add p2wpkh and p2wsh spending test and fixes to sign calls in wallet
* simplify wallet signing calls in cryptoengine
* add p2wpkh engine, add bip84 wallet

(Edit: this has been rebased on #205 now). First, a couple of notes to @undeath specifically: I realise that this may conflict with #205, and it'd be a bit "off" to not address that first. I'm happy to do that, but during the process of putting this together I realised there were a couple of things about the way `cryptoengine` worked that I wasn't happy with (the obvious: re-doing bitcoin signing logic; that's been addressed here). We can discuss the details, but I'll be reading those changes again next, and if I have to just rebase this after, so be it.

Second, to @undeath , please note there's some refactoring in the wallet classes, most particular is creating an intermediate class for purposed-bip32, so that 49 and 84 can avoid code reuse. I guess this part you'll be fine with, in particular it kind of shows off what's good about what you did in the first place here :) But just generally I'm sure you'll have some thoughts on the various changes there (get_script_code may be of interest, too).

Note that new [tests](https://github.com/JoinMarket-Org/joinmarket-clientserver/pull/268/files#diff-ecb9c3bba46f23ba646ddea4f68ea4e6R210) have been added specifically to check the ability to sign, and verify, p2wpkh and p2wsh type utxos; this is not because we have an urgent desire to start using these in Joinmarket, more that it feels like the various paths forward from here should all involve us being able to handle transactions using these different segwit variants, so we kind of need that infrastructure.

Additionally a [BIP84](https://github.com/bitcoin/bips/blob/master/bip-0084.mediawiki) wallet class has been created; it's an exact analogue of BIP49 but for p2wpkh instead of p2sh-p2wpkh as we currently use. So it switches to bech32 addresses instead of '3' p2sh addresses. Note that this has not been implemented anywhere else in the codebase; to get output like [this](https://0bin.net/paste/pInHlajEMDIyyrTH#UT0E-iexwO83KP8pgFI8XPmMCeP4IEUU8klolBltMZJ) you'd have to add a few hack-lines to the wallet creation code (but it's really just like 3 lines, which is cool :) ).

Obviously implementing a BIP84 wallet for users can, and probably will be done later, but it will involve us deciding what path forward we want with respect to Joinmarket wallets. One option is to have multiple wallets from one seed, on the different HD paths; another (not necessarily different) possibility might be having totally mixed-address coinjoins. Sounds complicated, but I'm really not sure. I guess the 'easy' option is to just migrate to p2wpkh at some point, which will be the most space-economical, but as we've observed historically, there are of course costs to migrating.

Finally to state the obvious there are a lot of little edits in tests and so on, some of it due to the new signing and verifying code in `secp256k1_transaction.py` and some of it due to some edits in the wallet and cryptoengine modules.

Also note that I spent some time carefully thinking about how the signature material is passed from maker to taker, and so that code has changed somewhat.

I also added quite a few additional comments; although I think we're still not commented enough in a lot of functions.

AdamISZ commented:
Fixed flake error; the other error is specific to python2.7.5 or lower afaict; I don't get it locally on python2.7.12.

Basically the first argument to struct.unpack is not allowed to be a unicode string on py2.7.5 or earlier (what travis uses); see discussion [here](https://github.com/celery/kombu/issues/622); I'm not sure of a way to make that work, any help would be appreciated on that.

jameshilliard commented on jmbitcoin/jmbitcoin/secp256k1_transaction.py:
```diff
@@ -388,57 +413,122 @@ def address_to_script(addr):
             hrp = 'tb'
         return bech32addr_encode(hrp=hrp, witver=witver,
-                                 witprog=[ord(x) for x in script[2:]])
+        witprog=struct.unpack('{}B'.format(len(script[2:])), script[2:]))
```
Try changing this so that the first argument to `struct.unpack` is bytes.

AdamISZ replied:
Yes, thanks; at first I tried `'b{}B'.format(...` but of course that wasn't viable, format requires a string; can I cast to bytes in a way that's 2/3 compatible? bytes()? I guess I should just figure it out myself but in the middle of something else, if you know off the top of your head, let me know, but don't put yourself out otherwise :)

jameshilliard replied:
Maybe just encode('ascii') the entire first argument to struct.unpack. Using format strings with struct.unpack seems strange here, it's not something I recall ever seeing before, maybe [struct](https://docs.python.org/3/library/struct.html) has a more suitable function?

AdamISZ commented:
Dropping a note here in case I forget. pubkey_to_scrip functions in cryptoengine for p2wpkh and p2sh wrapped p2wpkh must *not* allow uncompressed pubkeys as currently are allowed in _pubkey_to_script - this is a stipulation of BIP141,143.

AdamISZ commented:
Rebased after merge of #205 . There were a few changes to make as a result of that, also fixed some other errors cropping up from testing.

Some more testing will be needed, but this should be mostly done now.

undeath commented on jmclient/jmclient/maker.py:
```diff
@@ -139,14 +139,30 @@ def on_tx_received(self, nick, txhex, offerinfo):
-
         for index in our_inputs:
-            sigmsg = txs['ins'][index]['script']
+            sigmsg = unhexlify(txs['ins'][index]['script'])
```
`sign_tx` is called with non-hex data and the return data should be non-hex, too. If it's not that seems wrong.

AdamISZ replied:
It's not quite like that. I'm going to add a commit updating the comments to btc.serialize in particular, and also btc.sign, because the former especially describes what happens wrong.

serialize() is using *any* outpoint:hash field being hex in the input transaction dict as a flag to output hex serialization. sign() just obeys the encoding of the input *serialized* transaction; it accepts only a serialized tx, and outputs a hex re-serialization if input is hex, and binary if binary.

Yes this is a slight delta to previous (in particular the sign()) behaviour. I know we obviously we want to completely change this, but for now at least I'm going to try to make sure the comments are accurate.

AdamISZ replied:
Thanks for paying careful attention to this; in fact the comment was *still* wrong; I have updated in  . In short, jmbitcoin.sign() *always* returns hex as currently coded, hence wallet.sign_tx does also.

So overall: deserialize() respects encoding of input; serialize chooses hex if any hash input is hex; sign always returns hex (and only works with serialized form, cf other comments).

fivepiece commented on jmbitcoin/jmbitcoin/secp256k1_transaction.py:
```diff
@@ -598,6 +730,26 @@ def p2sh_p2wpkh_sign(tx, i, priv, amount, hashcode=SIGHASH_ALL, usenonce=None):
     txobj["ins"][i]["txinwitness"] = [sig, pub]
     return serialize(txobj)
 
+def p2wpkh_sign(tx, i, priv, amount, hashcode=SIGHASH_ALL, usenonce=None):
```
Can actually merge `p2sh_p2wpkh_sign` and `p2sh_p2wpkh` to avoid code re-use.  The sighash is the same, only difference is whether to place the p2wpkh witness program in scriptsig for p2sh-p2wpkh as redeemscript or leave it blank for bare p2wpkh

AdamISZ replied:
Excellent point, thanks. Done in https://github.com/JoinMarket-Org/joinmarket-clientserver/pull/268/commits/0b558247dd12971d842136724862672c3c204f24

AdamISZ commented on jmclient/jmclient/taker.py:
```diff
@@ -567,24 +567,48 @@ def on_sig(self, nick, sigb64):
+                # implicitly, as verify_tx_input used to, by reconstructing
+                # from the public key. For these cases, we can *assume* that
+                # the input is of type p2sh-p2wpkh
+                scriptCode = "76a914" + btc.hash160(unhexlify(ver_pub)) + "88ac"
```
Initially I fully agreed with this, and wanted to delegate it to the wallet, but then realised: we can't do that, because the wallet may not know how to get the scriptCode (if e.g. it's a LegacyWallet, see e.g. the mixed maker test in test_coinjoin), it's fundamentally a general bitcoin operation; we *can* put it in cryptoengine but that comes back to the question of whether we really do want a layer on top of the jmbitcoin package there (see my earlier comments; I'm ambivalent about it). So currently the above has been replaced with a call to `btc.pubkey_to_p2pkh_script` (which is the same as you'd get by calling wallet.get_script_code if your wallet knew this particular script).

undeath replied:
I must admit I'm not 100% sure myself how this should ideally be handled. Adding wrappers to the wallet/cryptoengine to do wallet-independent operations does not seem like good design. I think delegating to jmbitcoin here is the correct thing to do.

AdamISZ commented on jmclient/jmclient/wallet.py:
```diff
@@ -381,6 +383,19 @@ def script_to_addr(self, script):
+        """
+        path = self.script_to_path(script)
+        priv, engine = self._get_priv_from_path(path)
+        pub = btc.privkey_to_pubkey(priv, False)
```
Done in

AdamISZ commented on jmclient/jmclient/cryptoengine.py:
```diff
@@ -18,16 +18,6 @@
-P2WPKH_PRE = b'\x00\x14'
-
-
 def _pubkey_to_script(pubkey, script_pre, script_post=b''):
```
Agreed and done in 
I also thought about moving `detect_script_type` to jmbitcoin but decided not to bother for now, since the TYPE_ enum is specific to jmclient atm. Can be done of course.

Additional note: made a generic `is_valid_pubkey` method in secp*main; it checks as per your original `_pubkey_to_script` on valid pubkey types, but also addresses a point I mentioned above: has a flag to insist on a compressed serialization, and this gets used in the cryptoengine engines that are segwit style.

AdamISZ commented on jmclient/jmclient/cryptoengine.py:
```diff
@@ -278,62 +263,59 @@ def VBYTE(cls):
-
-        return tx
-
+        return btc.sign(btc.serialize(tx), index, privkey,
```
On this one, I definitely agree with your thinking - not good/logical to pass serializations instead of objects. But I think I'll leave this one alone, for now.

AdamISZ commented on jmbitcoin/jmbitcoin/secp256k1_transaction.py:
```diff
@@ -388,57 +413,122 @@ def address_to_script(addr):
+    """
+    if not isinstance(pub, bytes):
+        pub = binascii.unhexlify(pub)
+    return "0014" + hash160(pub)
```
Yes, good point, thanks. Done in 

Note that several binary and hex literals have been replaced with refs to the constants, but there is still one dangling "16" that I decided not to touch for now.

undeath commented on jmclient/jmclient/wallet.py:
```diff
@@ -331,7 +332,8 @@ def sign_tx(self, tx, scripts, **kwargs):
             path = self.script_to_path(script)
             privkey, engine = self._get_priv_from_path(path)
-            engine.sign_transaction(tx, index, privkey, amount, **kwargs)
+            tx = btc.deserialize(engine.sign_transaction(tx, index, privkey,
```
if `engine.sign_transaction` receives a deserialized tx I think it should also return a deserialized tx.

undeath commented on jmclient/jmclient/taker.py:
```diff
@@ -740,13 +766,7 @@ def self_sign(self):
-
-        self.latest_tx = btc.deserialize(hexlify(btc.serialize(tx_bin)).decode('ascii'))
-
+        self.latest_tx = self.wallet.sign_tx(self.latest_tx, our_inputs)
```
Does this work without the hexlify/unhexlify hack? We wouldn't want to accidentally end up with hex and non-hex data in the tx object.

AdamISZ replied:
As noted in the earlier comment [here](https://github.com/JoinMarket-Org/joinmarket-clientserver/pull/268#discussion_r244640376), because of the behaviour (now correctly documented) of jmbitcoin.sign, this will always come out hex encoded.

undeath commented on jmbitcoin/jmbitcoin/secp256k1_transaction.py:
```diff
@@ -388,57 +415,141 @@ def address_to_script(addr):
+    see BIP141 for definitions of 2 current valid scripts.
+    """
+    if not isinstance(script, bytes):
+        script = binascii.unhexlify(script)
```
Why do some is_*_script functions have a hex check and others don't? Ideally we would get rid of those checks and only handle non-hex data.

AdamISZ replied:
Good point. Occasionally there are calls directly from non-wallet code e.g. `is_native_segwit_script` is called by taker in a hex context. Most other calls from outside jmbitcoin are from the wallet code, which is all binary. I've added the above check to the other functions that can be called from outside the package, in

AdamISZ commented:
Cancel the previous comment - it was not ready because of something I had not yet tested - old takers are not compatible with new makers in that version; the maker must, unfortunately, send the witnessProgram specifically as the third field of sigmsg because the old takers use that and insert it into the scriptSig directly. Reverted that and altered the comments to correspond. Note that this probably implies there is no clever way to make existing code compatible with any update to a new wallet type (like BIP84), at least I think so. Will have to update that in a non-compatible way.

AdamISZ commented on jmbitcoin/jmbitcoin/secp256k1_transaction.py:
```diff
@@ -388,57 +415,145 @@ def address_to_script(addr):
             hrp = 'tb'
         return bech32addr_encode(hrp=hrp, witver=witver,
-                                 witprog=[ord(x) for x in script[2:]])
+        witprog=struct.unpack('{}B'.format(len(script[2:])).encode(
```
The update here is actually a bugfix; without it, Py3 makers fail to process bech32 destination addresses correctly.

AdamISZ commented:
So with that final reversion of the sig message contents, I am finding this working correctly in all modes.

Plan to merge it in a couple of days unless someone finds further issues.
